velox/connectors/hive/HiveDataSink.h

```diff
@@ -192,16 +192,96 @@ class HiveDataSink : public DataSink {
 
  private:
   std::unique_ptr<dwrf::Writer> createWriter(
-      const std::shared_ptr<const HiveWriterParameters>& writerParameters);
+      const std::shared_ptr<const HiveWriterParameters>& writerParameters)
+      const;
+
+  // Move the existing writers to the new indices, if the current input causes
+  // the partitionIdGenerator_ to rehash that changes the IDs of the previous
+  // partition values and the corresponding writers.
+  void moveWriters();
+
+  static RowVectorPtr makeWriterInput(
+      const RowVectorPtr& rawInput,
+      vector_size_t size,
+      const BufferPtr& index);
+
+  // Compute the count of rows as well as the concrete row indices for every
+  // partition ID based on the ID labeling of rowPartitionIds. Both rowCounts
+  // and rowIndices are indexed by partition ID. rowIndices[i] is a vector of
+  // all rows labeled with partition ID i.
+  static void computePartitionRowCountsAndIndices(
+      const raw_vector<uint64_t>& rowPartitionIds,
+      const SelectivityVector& activeRows,
+      uint64_t maxPartitionId,
+      memory::MemoryPool* FOLLY_NONNULL pool,
+      std::vector<vector_size_t>& rowCounts,
+      std::vector<BufferPtr>& rowIndices);
+
+  RowVectorPtr makeInputPartitionsVector(const RowVectorPtr& input) const;
+
+  std::vector<BaseVector::CopyRange> convertToIncrementalRanges(
+      std::vector<vector_size_t>& rowCounts,
+      std::vector<BufferPtr>& rowIndices) const;
+
+  static std::string getPartitionString(
+      const RowVectorPtr& partitionsString,
+      vector_size_t row);
 
   const RowTypePtr inputType_;
   const std::shared_ptr<const HiveInsertTableHandle> insertTableHandle_;
   const ConnectorQueryCtx* FOLLY_NONNULL connectorQueryCtx_;
   const std::shared_ptr<WriteProtocol> writeProtocol_;
-  // Parameters used by writers, and thus are tracked in the same order
+  const std::vector<column_index_t> partitionChannels_;
+  const RowTypePtr partitionsType_;
+  PartitionIdGenerator partitionIdGenerator_;
+
+  // Parameters used by writers, and thus are indexed in the same order
   // as the writers_ vector
   std::vector<std::shared_ptr<const HiveWriterParameters>> writerParameters_;
   std::vector<std::unique_ptr<dwrf::Writer>> writers_;
+  // Distinct partition values that writers write for, and thus are indexed in
+  // the same order as the writers_ vector. This RowVector is kept to regenerate
+  // IDs for the existing writers if new input makes the partitionIdGenerator
+  // rehash.
+  RowVectorPtr partitionsVector_;
+  uint32_t numPartitions;
+
+  // Below are structures updated when generate partition IDs for the current
+  // input. All are indexed by the row of input_.
+  RowVectorPtr input_;
+  SelectivityVector activeRows_;
+  raw_vector<uint64_t> partitionIds_;
 };
 
+#define PARTITION_TYPE_DISPATCH(TEMPLATE_FUNC, typeKind, ...)             \
```

Comment:
This macro is only used in the .cpp file. Can it be moved there?

It also looks very similar to VALUE_ID_TYPE_DISPATCH defined in VectorHasher.cpp. It would be nice to consolidate these two to avoid copy paste.

Reply:
Moving the macro to .cpp file. To consolidate this macro w/ existing macros, I think I can directly call VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH, to avoid moving VALUE_ID_TYPE_DISPATCH from .cpp to .h.